src/journal/compress.c

```diff
@@ -667,13 +677,266 @@ int decompress_stream_lz4(int in, int out, uint64_t max_bytes) {
         return -EPROTONOSUPPORT;
 #endif
 }
+#if HAVE_ZSTD
+typedef struct MmapAlloc {
+        void *alloc;
+        size_t size;
+} MmapAlloc;
+
+_nonnull_() static void free_mmap_allocp(MmapAlloc *p) {
+        if (p->size)
+                munmap(p->alloc, p->size);
+}
+
+static int zstd_ret_to_errno(size_t ret) {
+        static_assert(ZSTD_VERSION_NUMBER >= 10301,
+                      "<zstd_errors.h> is stable since zstd 1.3.1");
+
+        ZSTD_ErrorCode error = ZSTD_getErrorCode(ret);
+
+        switch (error) {
+        case ZSTD_error_dstSize_tooSmall:
+                return -ENOBUFS;
+        case ZSTD_error_memory_allocation:
+                return -ENOMEM;
+        default:
+                return -EBADMSG;
+        }
+}
+#endif
+
+int compress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeCCtxp) ZSTD_CCtx *cctx = NULL;
+        _cleanup_(free_mmap_allocp) MmapAlloc map_alloc = {NULL, 0};
+        size_t buffInSize, buffOutSize;
+        void *buffIn, *buffOut;
+        size_t z;
+        uint64_t left = max_bytes, in_bytes = 0;
+        uint64_t inputSize = 0;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        /* Create the context and buffers */
+        buffInSize = ALIGN_POWER2(ZSTD_CStreamInSize());
+        buffOutSize = ALIGN_POWER2(ZSTD_CStreamOutSize());
+        map_alloc.alloc = mmap(NULL, buffInSize + buffOutSize, PROT_READ | PROT_WRITE,
+                           MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
```

Comment:
I don't understand why `mmap` is used. Two `malloc()s`, one for `buffInSize`, and one for `buffOutSize` should work just as well. Those calls `ALIGN_POWER2` should not be necessary then either.

Reply:
I mentioned sparse files. Those would require us to search for filesystem blocks that are empty (zeros), and instead of writing out the data, to do a `lseek` over those blocks.
The idea was to get suitably aligned buffers (filesystem block size), easing up the logic. I am not so sure about that anyway.